[UX/Feature] Allowing an autostop hook that runs before the cluster autostop (#8412)
skypilot-org/skypilot: docs/source/reference/auto-stop.rst

   # Hard time limit: Stop after 10 minutes, regardless of running jobs or SSH sessions.
   sky autostop mycluster -i 10 --wait-for none

.. _auto-stop-hooks:

Autostop hooks
~~~~~~~~~~~~~~

To execute a script before autostopping, specify a hook in the autostop configuration.
The hook script runs on the remote cluster before the cluster is stopped or torn down.
This is useful for tasks like committing code, saving checkpoints, or performing cleanup operations.

.. code-block:: yaml

   resources:
     autostop:
       idle_minutes: 10
       hook: |
         cd my-code-base
         git add .
         git commit -m "Commit my code"
         git push

The hook script runs on the cluster and has access to the cluster's filesystem and environment variables.
If the hook script fails (non-zero exit code), the autostop process will still continue,
but a warning will be logged.

**Hook Timeout**

By default, autostop hooks have a **1-hour (3600 seconds) timeout**. If your hook

> takes longer than this, it will be terminated and autostop will proceed. You can

Reviewer: ```suggestion
takes longer than this, it will be killed and autostop will proceed. To
```

customize the timeout in your YAML configuration:

.. code-block:: yaml

   resources:
     autostop:
       idle_minutes: 10
       hook: |
         # Long-running backup operation
         tar -czf backup.tar.gz /large/dataset
         aws s3 cp backup.tar.gz s3://my-bucket/
       hook_timeout: 7200  # 2 hours in seconds

**Important Notes:**

- If the hook times out, autostop will still proceed after logging a warning
- The minimum timeout is 1 second

> - Consider the hook execution time when setting ``idle_minutes``

Reviewer: ```suggestion
- Hook execution will keep the cluster UP, occupying the resources. Be aware of that when setting ``idle_minutes``
```

> - Hook execution will keep the cluster UP, occupying the resources. Be aware of that when setting ``idle_minutes``

Reviewer 2: > keep the cluster UP

This is misleading since the status will be AUTOSTOPPING. How about "will keep the cluster from terminating while it runs"?

> Common use cases for autostop hooks:

Reviewer 2: I am a little hesitant to suggest these because most of these things are better accomplished by other mechanisms:
- git commit: not really needed because autostop will retain disk
- checkpoints/upload logs/w&b sync: should happen as part of a job
- slack notification: this one is really good

In general I'm a little wary of telling people to rely on this for saving code or data since there are a lot of cases where the hook might not get called (e.g. direct termination, termination on the cloud, cluster becomes unhealthy, etc)

Author: Sync with @Michaelvll for interactive use cases: e.g., when a `sky launched VM` is autostopped/terminated after a user SSHed in and made code/file changes, we need to commit/sync the DB before that happens.

So we need to keep these examples.

   # Hard time limit: Stop after 10 minutes, regardless of running jobs or SSH sessions.
   sky autostop mycluster -i 10 --wait-for none

Autostop hooks
~~~~~~~~~~~~~~

To execute a script before autostopping, specify a hook in the autostop configuration.
The hook script runs on the remote cluster before the cluster is stopped or torn down.
This is useful for tasks like committing code, saving checkpoints, or performing cleanup operations.

.. code-block:: yaml

   resources:
     autostop:
       idle_minutes: 10
       hook: |
         cd my-code-base
         git add .
         git commit -m "Commit my code"
         git push

The hook script has access to the cluster's filesystem and environment variables.
If the hook script fails (non-zero exit code), the autostop process will still continue,

> but a warning will be logged. The hook execution has a 1-hour timeout.

Reviewer: where is the warning logged to?

Common use cases for autostop hooks include:

- Committing and pushing code changes
- Saving model checkpoints to persistent storage
- Uploading logs or results to cloud storage
- Cleaning up temporary files

> - Sending notifications about the cluster shutdown

Author: <img width="1509" height="1698" alt="image" src="https://github.com/user-attachments/assets/89550545-7a0b-4a6e-b111-5d86b0502f84" />
Added

.. dropdown:: Saving model checkpoints to persistent storage

    .. code-block:: yaml

       resources:
         autostop:
           idle_minutes: 10
           hook: |
             # Save checkpoints to a mounted volume or cloud storage
             cp -r /workspace/checkpoints/* /mnt/persistent-storage/checkpoints/
             # Or upload to S3
             aws s3 sync /workspace/checkpoints/ s3://my-bucket/checkpoints/

.. dropdown:: Uploading logs or results to cloud storage

    .. code-block:: yaml

       resources:
         autostop:
           idle_minutes: 10
           hook: |
             # Upload logs to S3
             aws s3 sync /workspace/logs/ s3://my-bucket/logs/$(date +%Y%m%d)/
             # Or upload to GCS
             gsutil -m cp -r /workspace/results/ gs://my-bucket/results/$(date +%Y%m%d)/

.. dropdown:: Cleaning up temporary files

    .. code-block:: yaml

       resources:
         autostop:
           idle_minutes: 10
           hook: |
             # Remove temporary files to save space
             find /tmp -type f -mtime +1 -delete
             find /workspace/.cache -type f -mtime +7 -delete
             # Clean up Docker images if needed

>              docker system prune -af --volumes

Reviewer: I don't think this is needed.

Author: Done.
<img width="1947" height="1311" alt="image" src="https://github.com/user-attachments/assets/65abf897-bef4-4222-8f5f-4548502719d7" />

.. dropdown:: Sending notifications about the cluster shutdown

    .. code-block:: yaml

       resources:
         autostop:
           idle_minutes: 10
           hook: |
             # Send email notification
             echo "Cluster shutting down after idle period" | \
               mail -s "Cluster Autostop" user@example.com
             # Or send Slack notification via webhook
             curl -X POST -H 'Content-type: application/json' \
               --data '{"text":"Cluster shutting down after idle period"}' \

>                https://hooks.slack.com/services/YOUR/WEBHOOK/URL

Reviewer: is this the correct URL for slack?

Author: > `https://hooks.slack.com/services/T00000000/B00000000/XXXXXXXXXXXXXXXXXXXXXXXX`

Yes, see [slack doc](https://docs.slack.dev/messaging/sending-messages-using-incoming-webhooks/)
